### ISSUE TITLE
BUG: Deepredeff output is empty

### ISSUE THREAD

**Issue by darcyabjones** [ccdmb/predector#87]
@JWDebler reported that the deepredeff results are empty.
Some preliminary work by him suggests that deepredeff fails but doesn't raise an error so the program continues without those results.

I suspect it has something to do with the whole Keras thing.

**Comment by darcyabjones** [comment]
I wasn't able to reproduce this with my conda environment, so it might be specific to @JWDebler

**Comment by JWDebler** [comment]
I'm running it using the docker container though though. Shouldn't that
make it independent of my system?

**Comment by darcyabjones** [comment]
Ah ok thanks. I'll try running with docker and see.
Yeah it should make it independent.

Usually with containers the issue is something to do with permissions or maybe something got corrupted when i built it.
Give me a minute

**Comment by darcyabjones** [comment]
Yep, it's a permissions error.

```
WARNING: Your kernel does not support swap limit capabilities or the cgroup is not mounted. Memory limited without swap.                                                                                                                                                                                                                                                                                                                                                                                                                                                        # >>>>>>>>>>>>>>>>>>>>>> ERROR REPORT <<<<<<<<<<<<<<<<<<<<<<                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        Traceback (most recent call last):                                                                                                                                                                                                                                                        File "/opt/conda/lib/python3.9/site-packages/conda/exceptions.py", line 1080, in __call__                                                                                                                                                                                                 return func(*args, **kwargs)                                                                                                                                                                                                                                                          File "/opt/conda/lib/python3.9/site-packages/conda/cli/main.py", line 84, in _main                                                                                                                                                                                                        exit_code = do_call(args, p)                                                                                                                                                                                                                                                          File "/opt/conda/lib/python3.9/site-packages/conda/cli/conda_argparse.py", line 83, in do_call                                                                                                                                                                                            return getattr(module, func_name)(args, parser)                                                                                                                                                                                                                                       File "/opt/conda/lib/python3.9/site-packages/conda/cli/main_run.py", line 25, in execute                                                                                                                                                                                                  script_caller, command_args = wrap_subprocess_call(on_win, context.root_prefix, prefix,                                                                                                                                                                                               File "/opt/conda/lib/python3.9/site-packages/conda/utils.py", line 403, in wrap_subprocess_call                                                                                                                                                                                           with Utf8NamedTemporaryFile(mode='w', prefix=tmp_prefix, delete=False) as fh:                                                                                                                                                                                                         File "/opt/conda/lib/python3.9/site-packages/conda/auxlib/compat.py", line 88, in Utf8NamedTemporaryFile                                                                                                                                                                                  return NamedTemporaryFile(                                                                                                                                                                                                                                                            File "/opt/conda/lib/python3.9/tempfile.py", line 541, in NamedTemporaryFile                                                                                                                                                                                                              (fd, name) = _mkstemp_inner(dir, prefix, suffix, flags, output_type)
      File "/opt/conda/lib/python3.9/tempfile.py", line 251, in _mkstemp_inner
        fd = _os.open(file, flags, 0o600)
    PermissionError: [Errno 13] Permission denied: '/opt/conda/envs/predector/.tmpearazcnw'

`$ /opt/conda/bin/conda run --prefix /opt/conda/envs/predector --no-capture-output python -c import os; print(os.environ['PATH'])`

  environment variables:
                 CIO_TEST=<not set>
             CONDA_PREFIX=/opt/conda/envs/predector
               CONDA_ROOT=/opt/conda
                    CPATH=:/opt/conda/envs/predector/include
           CURL_CA_BUNDLE=<not set>
          LD_LIBRARY_PATH=:/opt/conda/envs/predector/lib
             LIBRARY_PATH=:/opt/conda/envs/predector/lib
                     PATH=/home/ubuntu/.nextflow/assets/ccdmb/predector/bin:/opt/conda/envs/pred
                          ector/bin:/opt/conda/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr
                          /bin:/sbin:/bin
               PYTHONPATH=/opt/conda/envs/predector/lib/python3.6/site-packages:
       REQUESTS_CA_BUNDLE=<not set>
            SSL_CERT_FILE=<not set>


     active environment : predector
    active env location : /opt/conda/envs/predector
       user config file : /.condarc
 populated config files :
          conda version : 4.11.0
    conda-build version : not installed
         python version : 3.9.7.final.0
       virtual packages : __linux=5.4.0=0
                          __glibc=2.31=0
                          __unix=0=0
                          __archspec=1=x86_64
       base environment : /opt/conda  (read only)
      conda av data dir : /opt/conda/etc/conda
  conda av metadata url : None
           channel URLs : https://repo.anaconda.com/pkgs/main/linux-64
                          https://repo.anaconda.com/pkgs/main/noarch
                          https://repo.anaconda.com/pkgs/r/linux-64
                          https://repo.anaconda.com/pkgs/r/noarch
          package cache : /opt/conda/pkgs
                          /.conda/pkgs
       envs directories : /.conda/envs
                          /opt/conda/envs
               platform : linux-64
             user-agent : conda/4.11.0 requests/2.27.1 CPython/3.9.7 Linux/5.4.0-117-generic debian/11 glibc/2.31
                UID:GID : 1000:1000
             netrc file : None
           offline mode : False


An unexpected error has occurred. Conda has prepared the above report.

Error in Sys.setenv(PATH = new_path) : wrong length for argument
Calls: main ... <Anonymous> -> python_config -> python_munge_path -> Sys.setenv
In addition: Warning message:
In system2(conda, c("run", in_env, run_args, shQuote(cmd), args),  :
  running command ''/opt/conda/bin/conda' run --prefix '/opt/conda/envs/predector' --no-capture-output 'python' -c "import os; print(os.environ['PATH'])"' had status 1
Execution halted
/opt/conda/envs/predector/lib/python3.9/site-packages/xgboost/compat.py:36: FutureWarning: pandas.Int64Index is deprecated and will be removed from pandas in a future version. Use pandas.Index with the appropriate dtype instead.
  from pandas import MultiIndex, Int64Index
```

The fact that it suppresses the exception is a bit worrying. I'm not sure if that's my wrapper of deepredeff

**Comment by darcyabjones** [comment]
Hey Johannes,

Sorry it's taken me a while to fix this.
I've put a fix in the docker container, would you be able to run the following to rebuild the docker container and see if it fixes your problem?

```
docker rmi predector/predector:1.2.6
docker rmi predector/predector-base:1.2.6
git clone https://github.com/ccdmb/predector.git && cd predector
docker build -t predector/predector-base:1.2.6 -f base.Dockerfile .
```

Then you can build the docker environment as usual with the `curl | install.sh` thing.
I think that should resolve your problem.

I haven't been able to track down why deepredeff is suddenly trying to write tempfiles to the conda directory, or why it fails without raising a detectable error. But at least it should be working

**Comment by darcyabjones** [comment]
So it turns out that with the update from python 3.6 to 3.9 deepredeff decided that it should try to install tensorflow itself again.

Internally deepredeff actually tries to create a conda environment, which always made it a real pain to deal with.
The problem that you were having was that it was trying to write to the conda environment (to install tensorflow), and install the particular version of tensorflow that they hard-coded into the package (apparently as a temporary fix for an old keras problem).
Anyway, it turned out to be much easier just to rewrite it in python, since it calls python anyway internally.
Deepredeff was the thing that made me scared to update software versions in predector.

**Comment by darcyabjones** [comment]
Implementation is here: https://github.com/darcyabjones/deepredeff-py

**Comment by darcyabjones** [comment]
Resolved with #90
